### ISSUE TITLE
Make unconfigured get properties return Dummies, as unconfigured methods do

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#156]
I would've expected (and I assume other users would too) unconfigured methods that return values and unconfigured get properties on fakes to return the same kinds of values. Not so!

If a getter's type is fakeable, we'll invoke the `AutoFakePropertyRule` (if not, then `DefaultReturnValueRule`). For a method, we fall back to `DefaultReturnValueRule`.

AutoFakePropertyRule will return a value obtained from `DefaultFakeAndDummyManager.CreateFake`, but DefaultReturnValueRule will return one from `DefaultFakeAndDummyManager.TryCreateDummy`.

In a lot of circumstances, the values will be the same, but if there's custom Dummy creation configured (either by a user or FakeItEasy itself), there will be different results. Let's look at an example!

```C#
[TestFixture]
public class DefaultPropertyMethodBehaviorTests
{
    public interface ILibrary
    {
        Movie FirstMovieOnShelfProperty { get; }
        Movie FirstMovieOnShelfMethod();
    }

    [Test]
    public void Should_return_same_type_for_unconfigured_method_and_property_getter()
    {
        var fakeLibrary = A.Fake<ILibrary>();

        var movieFromProperty = fakeLibrary.FirstMovieOnShelfProperty;
        var movieFromMethod = fakeLibrary.FirstMovieOnShelfMethod();

        Assert.That(movieFromProperty.GetType(), Is.EqualTo(movieFromMethod.GetType()));
    }

    public class Movie
    {
        public virtual string Title { get; set; }
    }

    public class DummyMovie: DummyDefinition<Movie>
    {
        protected override Movie CreateDummy()
        {
            return new Movie {Title = "Scott Pilgrim vs. the World"};
        }
    }
}
```

Gives us

```
Expected: <MethodPropertyTests.DefaultPropertyMethodBehaviorTests+Movie>
  But was:  <Castle.Proxies.MovieProxy>

at NUnit.Framework.Assert.That(Object actual, IResolveConstraint expression, String message, Object[] args)
at MethodPropertyTests.DefaultPropertyMethodBehaviorTests.Should_return_same_type_for_unconfigured_method_and_property_getter() in DefaultPropertyMethodBehaviorTests.cs: line 36 
```

The property gave us a Fake and the method gave us a Dummy (I think is preferable).

Note that I resorted to using a `DummyDefinition` to show the behaviour, but the same kind of thing happens when the return type is a `Task` or `Task<T>` (due to custom Tasky dummy handling added a while ago).

(I came across this when I was trying to write the the [Default fake behavior](../wiki/Default-fake-behavior) article and was experimenting. The bug is a big part of the reason it still says _"The actual rules for generating the return values are quite complicated, but here are some sample behaviours"_. It'd be convenient to be able to say "returns a Dummy".)

**Comment by adamralph** [comment]
As we are reaching late stages of preparation for our 2.0 release, it's worth considering any other breaking changes which we may or may not want to include.

Is it now clear what the proposed change is? Should it be part of 2.0?

**Comment by blairconrad** [comment]
_As I sit here_, it's not clear what proposed change @patrik-hagne and I would agree on. I will re-evaluate, along with #224 and #312, including my newfound understanding of `IFakeConfigurator`s. Then I'll see what I know.

However, as far as I know, I'm the only person who's noticed, let alone complained about, this behaviour, so if we end up delaying action on this, I don't mind. There's always 3.0!

**Comment by blairconrad** [comment]
It looks like #312 is going in the 2.0 release, but @patrik-hagne and I still have differences of opinion over how to treat this issue.
I propose leaving this issue out of 2.0. We can leave it open or, since I'm the only person who's ever complained about it, consider closing it (with the option of reopening if ever anyone else does run into a problem).

**Comment by adamralph** [comment]
I agree. Let's not hold up 2.0 for this.

**Comment by blairconrad** [comment]
Interestingly, recently a Stack Overflow question
[How to specify the class to use when faking a property?](http://stackoverflow.com/q/29282581/1199) was asked, and it seems that the "return Dummies from properties" behaviours would have accomplished exactly what the asker wanted.

**Comment by blairconrad** [comment]
There's been some offline chat about this issue, and #312 has been fixed, so the landscape has changed slightly. For the record, here are the highlights of the e-mail conversation the owners had over the last two months.

----

On 6 February 2015, I reraised this issue in e-mail to the owners, pressing for us to come to an agreement and try to get the issue fixed for the 2.0.0 release. I again pushed for having both unconfigured properties and methods return a Dummy.

----

On 7 February, after having me clarify some terminology, @adamralph indicated that he supported the change.

----

On 8 February, @patrik-hagne reminded us that a dummy is potentially any (non-null) value for a type, and that fakes can be dummies.
He expressed a preference for having unconfigured property getters 

1. Try creating a fake.
2. If that fails, try to create a dummy.
3. If that fails, use null.

(And to always return the same value, since implemented in #312 and to be released in 2.0.0-beta008, since nobody objected to that part.)

He solicited drawbacks to this approach.

----

Later on 8 February, @blairconrad summarized the current (and soon-to-be-current) Dummy creation logic:

1. Try constructing a dummy using a matching "dummy definition". (renamed "dummy factory" in 2.0.0-beta006)
2. If that fails, and the type is a Task, use the special Dummy Task creation mechanism. (added in 1.13)
3. If that fails, and the type is a Lazy, use the special Dummy Lazy creation mechanism. (added in 2.0.0-beta002)
4. If that fails, try to create a fake.
5. If that fails, and the type is a value type, create using Activator.CreateInstance
6. If that fails, and the type is not a Delegate, try to create using reflection, trying constructors with the longest parameter list first. (This is unlike when a Fake calls the base constructor - it tries the parameterless constructor first, then the ones with the longest argument list. But that is perhaps a discussion for another day.)

I thought that the drawbacks of combining this with preferring to return a Fake from a property are:

1. If the return type is a Task or a Lazy, the result will be a Fake Task or Fake Lazy, not Dummy versions.
    1. The Task won't appear to be completed, and waiting on it will never terminate.
    2. The Lazy won't appear to have a value. 
    3. Asking for the Lazy value will fail if the type does not have a parameterless constructor, unless the user performs additional configuration on the Lazy.
2. When a user has defined a Dummy factory for the returned type, they will get a Fake of the returned type instead of their configured Dummy, in contrast to the behaviour with an unconfigured method. This could be undesired and confusing behaviour.

I said that the drawback to preferring to return a Dummy are:

1. If a user has configured a dummy factory for the returned type (which doesn't return a Fake, although I'd expect most of them not to) and the user wants to configure the returned value when accessing it as a chained property, they will receive an error ("The specified object is not recognized as a fake object.").

I continued, stressing why I believe returning a Dummy is preferable:

1. the downsides identified in the "prefer to return Fakes from properties" case are more likely to occur, as all we need is for a property to return a type for which a dummy factory applies, instead of requiring a dummy factory to apply and for there to be configuration applied to the returned value
2. surprises arising from "prefer to return fakes from properties" are going to be harder for a user to diagnose, especially since I'd expect a user who's defined a dummy factory and who receives an error trying to configure a method on a type to which the factory applies to be a relatively sophisticated user, and the workaround should be clear to them

----

On 2 March, @adamralph agreed with my assessment, and expressed a desire to fix #156, but did not express an opinion regarding whether it should be fixed in 2.0 or not.

**Comment by adamralph** [comment]
I vote for doing this in 2.0

**Comment by blairconrad** [comment]
I agree. My inclination to wait waned after the StackOverflow question [How to specify the class to use when faking a property?][so] came in the other day and I basically had to explain how Dummies are used to create fakes and as return values from unconfigured methods, but not from unconfigured properties. Also when I realized that we could've provided a simpler answer for the poster had the properties and methods worked the same way.

[so]: http://stackoverflow.com/questions/29282581/how-to-specify-the-class-to-use-when-faking-a-property

**Comment by adamralph** [comment]
OK, let's go ahead and do it. Any objections should be voiced here before the 2.0 release.

**Comment by blairconrad** [comment]
We're close enough, I think, and work for 2.0 is dwindling. I'm starting.

**Comment by adamralph** [comment]
:+1:

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
